### PR TITLE
[megatron, vllm] feat: NVFP4 (W4A16) QAT training support via ModelOpt

### PR DESCRIPTION
**Pull request by jQizhang**
### What does this PR do?

This PR adds support for **NVFP4 (W4A16) Quantization-Aware Training (QAT)** in verl's Megatron training pipeline, with quantized weight transfer to the vLLM rollout engine for inference. The implementation leverages [NVIDIA ModelOpt](https://github.com/NVIDIA/Model-Optimizer) for quantization.

### Checklist Before Starting

- [x] Search for similar PRs. Paste at least one query link here: #5190 
- [x] Format the PR title as `[{modules}] {type}: {description}` (This will be checked by the CI)
  - `{modules}` include `fsdp`, `megatron`, `veomni`, `sglang`, `vllm`, `rollout`, `trainer`, `ci`, `training_utils`, `recipe`, `hardware`, `deployment`, `ray`, `worker`, `single_controller`, `misc`, `perf`, `model`, `algo`, `env`, `tool`, `ckpt`, `doc`, `data`, `cfg`, `reward`
  - If this PR involves multiple modules, separate them with `,` like `[megatron, fsdp, doc]`
  - `{type}` is in `feat`, `fix`, `refactor`, `chore`, `test`
  - If this PR breaks any API (CLI arguments, config, function signature, etc.), add `[BREAKING]` to the beginning of the title.
  - Example: `[BREAKING][fsdp, megatron] feat: dynamic batching`

### API and Usage Example

Enable NVFP4 QAT by adding the following to the actor Megatron config:

```yaml
qat:

  # Whether to enable QAT
  enable: false

  # Quantization mode: "w4a16" (weight-only)
  mode: "w4a16"

  # Quantization group size (NVFP4 requires 16)
  group_size: 16

  # Patterns to ignore (e.g., lm_head, embed_tokens)
  ignore_patterns:
    - "lm_head"
    - "*mlp.gate*"

  # Path to vLLM quantization config JSON file
  quantization_config_path: null
```

The training loop automatically:
1. Applies NVFP4 QAT quantizers to the Megatron actor model at initialization
2. Trains with simulated quantization (forward pass uses fake-quantized weights)
3. Before rollout, converts bf16 weights to NVFP4 format (packed uint8 + scales) and streams them to vLLM
4. vLLM loads the quantized weights using patched `ModelOptNvFp4LinearMethod` / `ModelOptNvFp4FusedMoE` for inference

### Experiments

- **Model**: Qwen3-30B-A3B (MoE)
- **Setup**: Training with Megatron, inference with vLLM
- **Quantization**: NVFP4 W4A16 via ModelOpt

image

verl-recipe PR: [verl-project/verl-recipe/pull/55](https://github.com/verl-project/verl-recipe/pull/55)

### Checklist Before Submitting

> [!IMPORTANT]
> Please check all the following items before requesting a review, otherwise the reviewer might deprioritize this PR for review.

- [x] Read the [Contribute Guide](https://github.com/volcengine/verl/blob/main/CONTRIBUTING.md).
- [x] Apply [pre-commit checks](https://github.com/volcengine/verl/blob/main/CONTRIBUTING.md#code-linting-and-formatting): `pre-commit install && pre-commit run --all-files --show-diff-on-failure --color=always`
- [ ] Add / Update [the documentation](https://github.com/volcengine/verl/tree/main/docs).
- [ ] Add unit or end-to-end test(s) to [the CI workflow](https://github.com/volcengine/verl/tree/main/.github/workflows) to cover all the code. If not feasible, explain why: ...
- [ ] Once your PR is ready for CI, send a message in [the `ci-request` channel](https://verl-project.slack.com/archives/C091TCESWB1) in [the `verl` Slack workspace](https://join.slack.com/t/verl-project/shared_invite/zt-3855yhg8g-CTkqXu~hKojPCmo7k_yXTQ). (If not accessible, please try [the Feishu group (飞书群)](https://applink.larkoffice.com/client/chat/chatter/add_by_link?link_token=772jd4f1-cd91-441e-a820-498c6614126a).)
- [x] If your PR is related to the `recipe` submodule, please also update the reference to the submodule commit via `git submodule update --remote` or `cd recipe && git pull origin main`.

### CODE REVIEW

File: verl/trainer/config/engine/megatron.yaml [hunk 1]
```diff
@@ -79,6 +72,12 @@ override_transformer_config:
   # Attention backend to use (flash,fused,unfused,local,auto). Defaults to auto in mcore, flash in verl
   attention_backend: flash
 
+# # Quantization method. None for no quantization, "nvfp4" for NVFP4 quantization
+quantization: null
+
+# Whether to enable Quantization-Aware Training (QAT). Default False.
+enable_qat: False
```
Comment (ISEEKYAN) [inline verl/trainer/config/engine/megatron.yaml]: better to use the same configs as in FSDP to avoid confusing the users

File: verl/workers/engine/megatron/transformer_impl.py [hunk 2]
```diff
@@ -173,6 +179,22 @@ def _build_tf_config(self):
             provider.moe_token_dispatcher_type = "alltoall"
             provider.moe_router_load_balancing_type = "none"
 
+            # Apply QAT: set quantization layer spec and patch Megatron-Bridge
```
Comment (ISEEKYAN) [inline verl/workers/engine/megatron/transformer_impl.py:199]: it will be more clear if we wrap the three parts of QAT code into three functions, also for readability.
and they can be shared with `verl/workers/megatron_workers.py`

Comment (jQizhang) [reply verl/workers/engine/megatron/transformer_impl.py:199]: Thanks for the feedback! I've split the QAT parts into functions to make it cleaner. Let me know if this looks good to you!
